pkgs/development/haskell-modules/generic-builder.nix

```
@@ -10,12 +10,81 @@
   runCommandCC,
   ghcWithHoogle,
   ghcWithPackages,
+  haskellLib,
+  iserv-proxy,
   nodejs,
+  writeShellScriptBin,
 }:
 
 let
   isCross = stdenv.buildPlatform != stdenv.hostPlatform;
 
+  crossSupport = rec {
+    emulator = stdenv.hostPlatform.emulator buildPackages;
+
+    # This is a silly workaround to avoid infinite recursion.
+    # We don't know which packages need template haskell, so we always need iserv-proxy for the host
+    # but then that package also has dependencies which would themselves need iserv-proxy.
+    # TODO: grab host iserv-proxy from the "original" package set without TH support, then have host TH support on the "real" package set.
+    iservProxyDependency =
+      pname:
+      builtins.elem pname [
+        "iserv-proxy"
+        "libiserv"
+        "network"
+      ];
+
+    hasBuiltinTH = stdenv.hostPlatform.isGhcjs;
+
+    canProxyTH =
+      pname:
+      lib.versionAtLeast ghc.version "9.6"
+      && stdenv.hostPlatform.emulatorAvailable buildPackages
+      && !(crossSupport.iservProxyDependency pname);
```

We're inside the `crossSupport = rec`, so:


```suggestion
      && !(iservProxyDependency pname);
```

---

`iservProxyDependency` is now gone